### DISCUSSION TITLE
5.0.0-beta.12

### DISCUSSION THREAD

**Discussion by adrianhall** [Azure/azure-mobile-apps#315] [General]
First full release of the complete set of libraries.

* SQLite store, with support for single table push/pull/purge.
* ms accuracy when storing datetimes.

<hr /><em>This discussion was created from the release <a href='https://github.com/Azure/azure-mobile-apps/releases/tag/5.0.0-beta.12'>5.0.0-beta.12</a>.</em>

**Comment by henda79** [comment]
Hey, I'm trying to get started with this, in particular the offline sync.

I've spent hours going through the sample code and docs but its very confusing as there is conflicting information. For example what is the interface IRemoteTable, google only returns 3 results and none of them are related.

The walkthough at:

https://docs.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/quickstarts/xamarin-forms/offline

mentions IMobileServiceSyncTable along with other interfaces and classes which I can't find any trace of in the source.

Any help, is appreciated.

**Reply by ideas2appsaustralia** [reply]
IMobileServiceSyncTable is an interface that can be accessed by inserting "Using Microsoft.WindowsAzure.MobileServices.Sync". I think you will find it in the AzureCloudTable<T> class

Are you using Adrians eBook? Its at [https://adrianhall.github.io/develop-mobile-apps-with-csharp-and-azure/] .From this and the github sample link in the ebook you have pretty well all you'll need to get going

**Reply by henda79** [reply]
@ideas2appsaustralia , thanks for your reply.

I never came across that link, so I will read up on that. However I am aware of the interface IMobileServiceSyncTable as there are a lot of resources around this, *but* the source on this repo does not use that interface, instead it uses IRemoteTable which is not documented anywhere.

I am not even sure what is recommended to use as it seems that MS come up with something new then stops support after only a few years and some of these articles are from 2020.

**Reply by ideas2appsaustralia** [reply]
the sample in Adrians eBook uses IMobileServiceSyncTable and is much more complete than the quick start. you will be fine working through the eBook. It will answer most of your questions

**Reply by adrianhall** [reply]
The samples that are "out there" all use the older version of the code (available at azure/azure-mobile-apps-net-server for about the next month before they are retired).  I'm currently working on the new docs to walk you through the new code.

**Reply by ideas2appsaustralia** [reply]
The code base to use with the eBook is at [https://github.com/adrianhall/develop-mobile-apps-with-csharp-and-azure/]

**Reply by henda79** [reply]
@adrianhall , thanks for the reply. I'm on a new project which requires offline sync so it seems pointless to integrate with something that is getting retired. However, I need to start implementing something.

I've had a look at the new code in Beta and managed to mangle an example by looking at the source. Do you recommend I spend the time to do this or use the old SDK if it's easy enough to refactor later when the docs are out ?

**Reply by henda79** [reply]
@adrianhall , I put together a small demo using the new SDK , but I can work out how to sync the offline with the online tables. The SyncContext class is declared as internal and it seems that has the methods required.

What am I missing ?

**Reply by adrianhall** [reply]
In the new SDK

1. Create an IOfflineStore:

    ``` csharp
    var store = new SqliteOfflineStore(connectionString);
    store.DefineTable<ModelClass>();
    ```

    Here, `connectionString` is a SQLite URI (i.e. something like `file://path/offline.db?mode=rwc`)

2. Create an options bundle and the client

   ``` csharp
   var options = new DatasyncClientOptions { OfflineStore = store };
   var client = new DatasyncClient(endpoint, options);
   ```

3. Initialize the store

    ``` csharp
    await client.InitializeOfflineStoreAsync();
    ```

Hope this helps!

**Reply by henda79** [reply]
Thanks @adrianhall, thanks for updating the docs for the new version of the SDK, it makes life that bit easier.

The part I'm struggling with, is syncing of the offline and online tables. I've looked at the docs https://github.com/Azure/azure-mobile-apps/wiki/How-to-use-the-.NET-Client-SDK#syncing-an-offline-table and implemented the code provided.

How do I get the offlineTables KVP ?

**Reply by adrianhall** [reply]
I don't understand that last question.  A typical push/pull synchronization would look like this:

``` csharp
foreach (var table in tableList) {
   await table.PushAsync();
   await table.PullAsync();
}
```

Admittedly, there are some rough edges there.  I'd be interested in what you were thinking?  What's the ideal code for you?

**Reply by henda79** [reply]
Sorry Adrian, I should have been clearer.

In the help docs the sample iterates over a offlineTables.Values collection.

```
foreach (var offlineTable in offlineTables.Values)
{
    await offlineTable.PushItemsAsync();
    await offlineTable.PullItemsAsync("", options);
}
```

I don't see any collection I can use in the SDK to iterate over, however there is an private field that keeps a track of defined tables which could be exposed ? 

```
  public class OfflineSQLiteStore : AbstractOfflineStore
  {
    private readonly Dictionary<string, TableDefinition> tableMap = new Dictionary<string, TableDefinition>((IEqualityComparer<string>) StringComparer.OrdinalIgnoreCase);
```

I know it would be very simple to create another collection but this seems like it goes against best practices.

Thanks.

**Reply by adrianhall** [reply]
The OfflineSQLiteStore is an implementation (eventually) of IOfflineStore, which is what you need to refer to when looking for definitions.  Basically, I am not assuming a specific implementation.  So you don't know how the data is being persisted.  It could, for example, all be stored as JSON blobs on a filesystem).  SQLite is a concrete implementation (and one we know and love).

However, good point - I'm filing an issue to "push all tables" on the client level.  That will be in the next release.

**Reply by henda79** [reply]
So what do you recommend to get the list of offline tables for now ?

**Reply by adrianhall** [reply]
Create the list yourself.

**Reply by henda79** [reply]
Ok, I was asking as there is an internal list which is added to when DefineTable is called.

**Reply by adrianhall** [reply]
As I said - that's private - either way, I'd need to do a release to get you the information.  Once we release the next beta, your problem will be solved.

**Reply by henda79** [reply]
ok, that sounds great. I'll hang back on the sync for a while. Is there a road map to get this to RC ?

Thanks for your help !

**Reply by adrianhall** [reply]
I'll do a respin of the beta on Friday and call that an RC.  I think we are there.
